File: src/hilbert/nodes/synthesis.py
"""Synthesis node for Hilbert.

Pipeline:
  1. Abstract compression — extract top-3 sentences per abstract by
     cosine similarity to the query embedding (no LLM call).
  2. Chunked synthesis — split papers into batches of CHUNK_SIZE and
     run synthesis LLM calls concurrently via asyncio.gather.
  3. Finding deduplication — embed all claims, merge any pair with
     cosine similarity > FINDING_DEDUP_THRESHOLD.
"""

import asyncio
import re
import uuid
from typing import List

from hilbert.llm import get_client, get_synthesis_prompt
from hilbert.llm.utils import parse_json_list
from hilbert.models import Finding, Paper
from hilbert.sources import cosine_similarity, get_embedding_client
from hilbert.state.research import ResearchState

_CHUNK_SIZE = 4               # papers per synthesis LLM call
_FINDING_DEDUP_THRESHOLD = 0.88   # claim similarity above which findings are merged
_TOP_SENTENCES = 3            # sentences to keep per abstract after compression


# ---------------------------------------------------------------------------
# Step 1 — Abstract compression
# ---------------------------------------------------------------------------

def _split_sentences(text: str) -> List[str]:
    """Split text into sentences using punctuation heuristics."""
    sentences = re.split(r'(?<=[.!?])\s+', text.strip())
    return [s.strip() for s in sentences if len(s.strip()) > 20]


async def _compress_abstracts(papers: List[Paper], query: str) -> List[dict]:
    """Return paper_data dicts with abstracts compressed to top-N sentences."""
    try:
        client = get_embedding_client()
        query_emb = (await client.embed_texts([query]))[0]

        compressed = []
        for paper in papers:
            sentences = _split_sentences(paper.abstract or "")
            if len(sentences) <= _TOP_SENTENCES:
                compressed.append({
                    "paper_id": paper.paper_id,
                    "title": paper.title,
                    "abstract": paper.abstract,
                })
                continue

            sent_embs = await client.embed_texts(sentences)
            scored = sorted(
                zip(sentences, sent_embs),
                key=lambda t: cosine_similarity(t[1], query_emb),
                reverse=True,
            )
            top_sentences = [s for s, _ in scored[:_TOP_SENTENCES]]
            compressed.append({
                "paper_id": paper.paper_id,
                "title": paper.title,
                "abstract": " ".join(top_sentences),
            })
        return compressed

    except Exception:
        # Fallback: return uncompressed
        return [
            {"paper_id": p.paper_id, "title": p.title, "abstract": p.abstract}
            for p in papers
        ]


# ---------------------------------------------------------------------------
# Step 2 — Chunked synthesis
# ---------------------------------------------------------------------------

async def _synthesise_chunk(query: str, chunk: List[dict]) -> List[Finding]:
    """Run one synthesis LLM call for a chunk of papers."""
    system_prompt, user_prompt = get_synthesis_prompt(query, chunk)
    try:
        client = get_client()
        content = await client.complete_text(
            prompt=user_prompt,
            system_prompt=system_prompt,
        )
        findings_data = parse_json_list(content)
        return [
            Finding(
                finding_id=f"finding-{uuid.uuid4().hex[:8]}",
                claim=fd.get("claim", ""),
                source_paper_id=fd.get("source_paper_id", ""),
                evidence_text=fd.get("evidence_text", ""),
            )
            for fd in findings_data
            if fd.get("claim")
        ]
    except Exception:
        return []


# ---------------------------------------------------------------------------
# Step 3 — Finding deduplication via claim embeddings
# ---------------------------------------------------------------------------

async def _dedup_findings(findings: List[Finding]) -> List[Finding]:
    """Merge near-duplicate findings (claim similarity > threshold).

    Keeps the finding with the higher confidence; concatenates evidence
    texts so no supporting detail is lost.
    """
    if len(findings) < 2:
        return findings

    try:
        client = get_embedding_client()
        claim_embs = await client.embed_texts([f.claim for f in findings])
    except Exception:
        return findings

    n = len(findings)
    merged_into: List[int] = list(range(n))  # union-find (flat)

    for i in range(n):
        for j in range(i + 1, n):
            if merged_into[j] != j:
                continue
            if cosine_similarity(claim_embs[i], claim_embs[j]) >= _FINDING_DEDUP_THRESHOLD:
                # Merge j into i — keep whichever has higher confidence
                if findings[j].confidence > findings[i].confidence:
                    findings[i].confidence = findings[j].confidence
                    findings[i].claim = findings[j].claim
                findings[i].evidence_text = (
                    findings[i].evidence_text + " " + findings[j].evidence_text
                ).strip()
                merged_into[j] = i

    return [f for i, f in enumerate(findings) if merged_into[i] == i]


# ---------------------------------------------------------------------------
# Main node
# ---------------------------------------------------------------------------

async def synthesis_node(state: ResearchState) -> dict:
    """Compress abstracts, chunk-synthesise in parallel, dedup findings."""
    query = state["query"]
    papers = state.get("papers", [])

    callback = state.get("progress_callback")
    if callback:
        callback("synthesis", {"papers": len(papers)})

    if not papers:
        return {"findings": [], "status": "verifying"}

    # Step 1: compress abstracts
    paper_data = await _compress_abstracts(papers, query)

    # Step 2: chunked parallel synthesis
    chunks = [
        paper_data[i: i + _CHUNK_SIZE]
        for i in range(0, len(paper_data), _CHUNK_SIZE)
    ]
    chunk_results = await asyncio.gather(
        *[_synthesise_chunk(query, chunk) for chunk in chunks],
        return_exceptions=True,
    )
    all_findings: List[Finding] = []
    for result in chunk_results:
        if isinstance(result, list):
            all_findings.extend(result)

    # Step 3: dedup by claim similarity
    findings = await _dedup_findings(all_findings)

    return {
        "findings": findings,
        "status": "verifying",
    }


def create_synthesis_node():
    """Create synthesis node function."""
    return synthesis_node

File: cli/src/core/nodes/synthesis.ts
import { getClient } from "../llm/client";
import { getSynthesisPrompt } from "../llm/prompts";
import { parseJsonList } from "../llm/utils";
import { getEmbeddingClient, cosineSimilarity } from "../sources/embeddings";
import { FINDING_DEDUP_THRESHOLD, CHUNK_SIZE } from "../constants";
import type { ResearchState } from "../state";
import type { Paper } from "../models/paper";
import type { Finding } from "../models/finding";

const TOP_SENTENCES = 3;

function splitSentences(text: string): string[] {
  const sentences = text.trim().split(/(?<=[.!?])\s+/);
  return sentences.map((s) => s.trim()).filter((s) => s.length > 20);
}

async function compressAbstracts(papers: Paper[], query: string): Promise<{ paper_id: string; title: string; abstract: string }[]> {
  try {
    const client = getEmbeddingClient();
    const queryEmbeds = await client.embedTexts([query]);
    const queryEmb = queryEmbeds[0];

    const compressed: { paper_id: string; title: string; abstract: string }[] = [];

    for (const paper of papers) {
      const sentences = splitSentences(paper.abstract || "");
      if (sentences.length <= TOP_SENTENCES) {
        compressed.push({
          paper_id: paper.paper_id,
          title: paper.title,
          abstract: paper.abstract,
        });
        continue;
      }

      const sentEmbeds = await client.embedTexts(sentences);
      const scored = sentences
        .map((s, i) => ({ text: s, emb: sentEmbeds[i] }))
        .sort((a, b) => cosineSimilarity(b.emb, queryEmb) - cosineSimilarity(a.emb, queryEmb));

      const topSentences = scored.slice(0, TOP_SENTENCES).map((s) => s.text);
      compressed.push({
        paper_id: paper.paper_id,
        title: paper.title,
        abstract: topSentences.join(" "),
      });
    }

    return compressed;
  } catch {
    return papers.map((p) => ({
      paper_id: p.paper_id,
      title: p.title,
      abstract: p.abstract,
    }));
  }
}

async function synthesiseChunk(
  query: string,
  chunk: { paper_id: string; title: string; abstract: string }[]
): Promise<Finding[]> {
  const [systemPrompt, userPrompt] = getSynthesisPrompt(query, chunk);
  try {
    const client = getClient();
    const content = await client.completeText(userPrompt, systemPrompt);
    const findingsData = parseJsonList(content);

    return findingsData
      .filter((fd) => (fd as Record<string, unknown>).claim)
      .map((fd) => {
        const f = fd as Record<string, unknown>;
        return {
          finding_id: `finding-${Math.random().toString(36).slice(2, 10)}`,
          claim: f.claim as string,
          source_paper_id: f.source_paper_id as string,
          evidence_text: f.evidence_text as string,
          confidence: 0,
          is_verified: false,
          created_at: new Date().toISOString(),
        } as Finding;
      });
  } catch {
    return [];
  }
}

async function dedupFindings(findings: Finding[]): Promise<Finding[]> {
  if (findings.length < 2) return findings;

  try {
    const client = getEmbeddingClient();
    const claimEmbeds = await client.embedTexts(findings.map((f) => f.claim));

    const n = findings.length;
    const mergedInto = Array.from({ length: n }, (_, i) => i);

    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        if (mergedInto[j] !== j) continue;
        if (cosineSimilarity(claimEmbeds[i], claimEmbeds[j]) >= FINDING_DEDUP_THRESHOLD) {
          if (findings[j].confidence > findings[i].confidence) {
            findings[i].confidence = findings[j].confidence;
            findings[i].claim = findings[j].claim;
          }
          findings[i].evidence_text = [findings[i].evidence_text, findings[j].evidence_text]
            .filter(Boolean)
            .join(" ")
            .trim();
          mergedInto[j] = i;
        }
      }
    }

    return findings.filter((_, i) => mergedInto[i] === i);
  } catch {
    return findings;
  }
}

export async function synthesisNode(state: ResearchState): Promise<Partial<ResearchState>> {
  const { query, papers } = state;

  const callback = state.progressCallback;
  if (callback) {
    callback("synthesis", { papers: papers.length });
  }

  if (!papers.length) {
    return { findings: [], status: "verifying" as const };
  }

  const paperData = await compressAbstracts(papers, query);

  const chunks: { paper_id: string; title: string; abstract: string }[][] = [];
  for (let i = 0; i < paperData.length; i += CHUNK_SIZE) {
    chunks.push(paperData.slice(i, i + CHUNK_SIZE));
  }

  const chunkResults = await Promise.all(
    chunks.map((chunk) => synthesiseChunk(query, chunk))
  );

  const allFindings: Finding[] = chunkResults.flat();

  const findings = await dedupFindings(allFindings);

  return {
    findings,
    status: "verifying" as const,
  };
}

export function createSynthesisNode() {
  return synthesisNode;
}